### DISCUSSION TITLE
Adding Bloch boundary condition in moose

### DISCUSSION THREAD

**Discussion by adhithya96** [idaholab/moose#26854] [Q&A Modules: General]
I want to add bloch type boundary condition for a wave propagation problem that I am solving. The  boundary condition is applied on both the displacement and stress (gradient of displacement). There are two files I found in moose AddPeriodicBCAction.c and FunctionPeriodicBoundary.c. Can someone help me with how to proceed?

**Comment by adhithya96** [comment]
I was trying to implement the suggestion in this #16419 by @crswong888. Ran into a problem of using complex number to signify the complex number  i in $`\exp(ikx)`$ in the bloch periodic boundary condition. Any ideas how to solve this problem ?
Basically incorporating the bloch periodic boundary condition i.e.
$`u_2 = u_1 * \exp(ikx)`$ 
using penalty formulation.

**Reply by GiudGiud** [reply]
Hello

You ll want to split your equation in the real and imaginary part.
Are u2 and u1 both complex numbers?

**Reply by adhithya96** [reply]
No. So I'm solving an eigenvalue problem to get dispersion curves. The displacements will get eliminated. The unknowns are wave number and eigenfrequency 
$K - \omega^2 M = 0$

Thanks for the quick reply. I am new to moose but the community is very helpful

**Reply by GiudGiud** [reply]
Can you split this K - w2 M  equation in its real and imaginary part?

**Comment by adhithya96** [comment]
Sorry for the late reply. I wrote a little derivation for the same. ABAQUS uses multipoint constraint to enforce bloch boundary condition. 
```math
\begin{equation}
    \begin{aligned}
        u(x +  h_i)  &= u(x)\text{exp}(i{k.h_i})\\
        \sigma(x +  h_i)  &= \sigma(x)\text{exp}(i{k.h_i})\\
    \end{aligned}
\end{equation}
 ```
```math
\begin{equation}
\begin{aligned}
    u(x) &=u^{Re}(x) + u^{Im}(x)\\
    \sigma(x) &= \sigma^{Re}(x) + i\sigma^{Im}(x)\\
\end{aligned}
\end{equation}
```
Substituting the second equation in first,
```math
\begin{equation}
\begin{aligned}
    u^{Re}(x + h) &= u^{Re}(x)cos(ka) - u^{Im}(x)sin(ka)\\
    u^{Im}(x + h) &= u^{Re}(x)sin(ka) + u^{Im}(x)cos(ka)\\
\end{aligned}
\end{equation}
```
and
```math
\begin{equation}
    \begin{aligned}
        \sigma^{Re}(x + h) &= \sigma^{Im}(x)sin(ka) - u^{Re}(x)cos(ka)\\
        \sigma^{Im}(x + h) &= -\sigma^{Re}(x)sin(ka) - \sigma^{Im}(x)cos(ka)\\
    \end{aligned}
\end{equation}
```
I am using two variables with same set of kernels for both i.e. MatDiffusion and MassEigenKernel . I am using LinearNodalConstraint for the displacement equation. For the stress equation I am not sure how to go about it. This where I am stuck now.

**Reply by GiudGiud** [reply]
once you have the stress in real & imaginary part, you have to add this stress term in the solid mechanics equation.

Are you planning to use the solid mechanics module? Or implement every equation yourself

**Reply by adhithya96** [reply]
No this is like a traction boundary condition. Shouldn't I use some form of Neumann boundary condition ?
The problem is this 
```math
\begin{equation}
\begin{aligned}
    \nabla . (C(x)  : \nabla_s u(x))  + \omega^2\rho(x)u(x)  &= 0\\
    u(x +  h_i)  &= u(x)\text{exp}(ik.h_i)\\
    \sigma(x + h_i). n(x + h_i) &= \sigma(x).n(x)\text{exp}(ik.h_i)
\end{aligned}
\end{equation}
```
I am solving a simplified 1-D problem where the $C(x)$ term becomes $E$. I am using MatDiffusion kernel for this and for the other term MassEigenKernel. The two equations below are the boundary conditions.

**Reply by GiudGiud** [reply]
I dont see sigma in your first equation?
Does it come in through C?

**Reply by adhithya96** [reply]
The equilibrium equation is simplified by substituting the constitutive relation $\sigma = C\epsilon$

**Reply by GiudGiud** [reply]
ok I see now.
is epsilon related to u ?
You have two complex variables, you ll likely need two (complex) equations to have a well posed problem. 

The first equation can be solved as two real equations with the right combination of kernels for each. 
The boundary conditions can either be written as periodic BC if the size of the domain is right on the period (2 pi / k )?

**Reply by adhithya96** [reply]
$\epsilon = 1/2(\nabla u + \nabla u^T)$
Regarding the complex variables, in the literature they are using the same mesh consdiering uim and ur as separate components. 
PeriodicBC won't work because we need to vary the value of k within the first Briollouin zone which can be like $[0,2\pi/a]$ to obtain omega for each. The goal is get the dispersion curves which is $\omega$ vs $k$ graph. PeriodicBC supports only $u(x + h_i) = u(x)$

**Reply by GiudGiud** [reply]
ok epsilon is strain. 
>  in the literature they are using the same mesh consdiering uim and ur as separate components.

ok good, we can do that.
ok no periodic BC.

How big is the solve going to be? I have a solution using functors, but it's not going to work in parallel without a lot more work.

**Comment by adhithya96** [comment]
Oh that's great. Parallel implementation won't be necessary. I will be using only 2-D mesh.

**Reply by adhithya96** [reply]
`[Mesh]
    [First]
        type = GeneratedMeshGenerator
        elem_type = EDGE
        dim =  1
        xmin = 0
        xmax = 10
        nx = 10
    []
[]
  
[Variables]
    [ur]
        order = FIRST
        family = LAGRANGE
    []
    [uim]
        order = FIRST
        family = LAGRANGE
    []
[]

[Kernels]
    [diff1]
        type = Diffusion
        variable = ur
    []
    [diff2]
        type = Diffusion
        variable = uim
    []
    [eigen1]
        type = MassEigenKernel
        variable = ur
    []
    [eigen2]
        type = MassEigenKernel
        variable = uim
    []
[]

[BCs]

[]


[Constraints]
    [./disp_x_1]
        type = LinearNodalConstraint
        variable = ur  
        variable_secondary = ur uim
        primary = left
        weights = '0.7071 -0.7071'
        secondary_node_set = right
        penalty = 1e8
        formulation = penalty
    [../]
    [./disp_x_1]
        type = LinearNodalConstraint
        variable = uim  
        variable_secondary = ur uim
        primary = left
        weights = '-0.7071 0.7071'
        secondary_node_set = right
        penalty = 1e8
        formulation = penalty
    [../]

[]
  
[Executioner]
    type = Eigenvalue
    eigen_problem_type = gen_non_hermitian
    which_eigen_pairs = SMALLEST_MAGNITUDE
    n_eigen_pairs = 5
    n_basis_vectors = 18
    solve_type = jacobi_davidson
    petsc_options = '-eps_view'
[]
  
[Outputs]
    exodus = true
[]'

**Reply by adhithya96** [reply]
This is the inp file I wrote for the case so far.

**Reply by GiudGiud** [reply]
ok so starting from the FunctorDirichletBC, you create new BCs and modify mainly this routine. There are some other changes to validParams and the constructor to do as well.
```
ADReal
FunctorDirichletBC::computeQpValue()
{
  // Rough pseudo codes, you ll need to get the syntax right, and add parameters (in validParam) to make it work
  // Get point locator
  const auto pl = _mesh.getPointLocator();
  // Locate node on the other side of the geometry
  // the translation will be a user parameter
  new_node = pl.locate_node(_current_node + _translation_vector);

  if (!new_node)
    mooseError("Did not find the opposite side value);

  const Moose::NodeArg space_arg = {new_node, Moose::INVALID_BLOCK_ID};


  const Moose::StateArg time_arg = Moose::currentState();
  return _coef(space_arg, time_arg) * _functor(space_arg, time_arg) * A exp(.....);
  // A and the exp depend on whether you are doing the real or imaginary part
}
```

then recompile and use it in the input

**Reply by adhithya96** [reply]
This is for the bloch boundary condition for displacement right ? But displacement boundary condition I can  apply using LinearNodalConstraint. It is the traction boundary condition where I am stuck.

**Reply by adhithya96** [reply]
Ok. will look into FunctorNeumannBC also ang get back. Thanks!

**Reply by GiudGiud** [reply]
ah good point. I lost track.
Yes if you can formulate it as a FunctorNeumannBC that could work.

**Comment by adhithya96** [comment]
https://github.com/adhithya96/Static1D
I tried out what you had suggested. BlochDirichletBC.C and BlochDirichletBC.h. (I couldn't upload all the files. Have to set up the github repository locally. will do it in a few days). I got these errors. 

'const class std::unique_ptr<libMesh::PointLocatorBase>' has no member named 'locate_node'
   54 |   const auto new_node = pl.locate_node(_current_node + _translation_vector);
      |                            ^~~~~~~~~~~
/home/adhithyar/Documents/Project/static1d/src/BlochDirichletBC.C:54:56: error: '_translation_vector' was not declared in this scope
   54 |   const auto new_node = pl.locate_node(_current_node + _translation_vector);

Couple of questions I have. 
1. What does the _functor and _coeff do ?
2. I was also trying to write a similar code for Neumann case. How do I mention the gradient of u on the one side of periodic boundary condition ? 
3. Is there anyway to learn how the code is written ?

**Reply by GiudGiud** [reply]
yes this was pseudo code. You have to find the actual names of the routines
see the doxygen
https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1PointLocatorBase.html

you need to add `_translation_vector` to the header too

1) Functor and coef can be evaluated anywhere on the mesh and provide the local field value
2) From a functor, you can get the gradient with `_functor.gradient(r,t)`
3) Use the doxygen. Or the VSCode integration that's also super helpful. And look at other objects in the folders

**Reply by adhithya96** [reply]
After going through the doxygen, I am even more confused. I get the error _curren_node is a const Node* but locate_node takes libMesh::point &  as the argument. I couldn't find any way to get  the  libMesh::point data from _mesh. The question I have is how is MOOSE able to convert the boundary condition specified by lets' say 
```
[BCs]
    [fixed1]
        type = DirichletBC
        variable = u
        boundary = 'boundary'
        value = 0
    []
[]
```
In the code  we don't write for every point in the specified boundary.

**Reply by GiudGiud** [reply]
A node can be caster to a point automatically. But a node pointer is not.

Just do *_current_node to get a Node and it can be used in lieu of a Point

**Reply by GiudGiud** [reply]
The dirichlet BC is applied on every node that is part of the 'nodeset' boundary

**Comment by adhithya96** [comment]
@GiudGiud  
This worked! I have implemented the bloch boundary condition for imaginary and real part of displacement. However, when I run the input file, I am getting the following error 
```
Invalid NodalBC for eigenvalue problems, please use homogeneous (array) Dirichlet
```
I've added all the files in a github repository - [Static1d](https://github.com/adhithya96/Static1D)
Any suggestions on how to proceed ?

**Reply by GiudGiud** [reply]
Are you using array variables?

**Reply by adhithya96** [reply]
I am using liMesh::Point to create  the translation vector
```
ADReal
BlochDirichletBCReal::computeQpValue()
{
  //Get point  locator  
  const auto pl = _mesh.getPointLocator();
  //Locate node on the other side of the geometry
  //The translation vector will be a user parameter
  libMesh::Point _translation_vec= libMesh::Point(_lattice_vec, 0.0, 0.0);
  const auto new_node = pl->locate_node(*_current_node + _translation_vec);


  if(!new_node)
    mooseError("Did not find the opposite side value");
  
  const Moose::NodeArg space_arg = {new_node, Moose::INVALID_BLOCK_ID};
  
  const Moose::StateArg time_arg = Moose::currentState();

  return _coef(space_arg, time_arg) * (_ur(space_arg, time_arg) * cos (_wave_num * _lattice_vec)  -  _uim(space_arg, time_arg) * sin (_wave_num * _lattice_vec));

}
```
It's the same for the imaginary part except for the return statement.

**Reply by GiudGiud** [reply]
do you have an input I can check out? This error is complaining about the base class of the BC. It does not seem to fit either the variable type or the solve type

**Reply by adhithya96** [reply]
[Static1D](https://github.com/adhithya96/Static1D)
This link has the file static1D_periodic_eigen.i which is what  I am using. I have also put the files needed  to run the input file. This is  how the input file looks like
```
[Mesh]
    [First]
        type = GeneratedMeshGenerator
        elem_type = EDGE
        dim =  1
        xmin = 0
        xmax = 10
        nx = 10
    []
[]
  
[Variables]
    [ur]
        order = FIRST
        family = LAGRANGE
    []
    [uim]
        order = FIRST
        family = LAGRANGE
    []
[]

[Kernels]
    [diff1]
        type = MatDiffusion
        variable = ur
        diffusivity = diffusivity 
    []
    [diff2]
        type = MatDiffusion
        variable = uim
        diffusivity = diffusivity 
    []
    [eigen1]
        type = CoefReaction
        variable = ur
        coefficient = -1.0
        extra_vector_tags = 'eigen'
    []
    [eigen2]
        type = CoefReaction
        variable = uim
        coefficient = -1.0
        extra_vector_tags = 'eigen'
    []
[]

[BCs]
    [Blochur]
        type = BlochDirichletBCReal
        variable = ur
        ur = 'ur'
        uim = 'uim'
        lattice_length = 10.0
        wave_number = 1.732
        boundary = 'left'
    []
    [Blochuim]
        type = BlochDirichletBCImag
        variable = uim
        ur = 'ur'
        uim = 'uim'
        lattice_length = 10
        wave_number = 1.732
        boundary ='left'
    []
[]

[Materials]
    [nm]
      type = GenericConstantMaterial
      prop_names = 'diffusivity'
      prop_values = 0.333333333333333333
    []
[]

[Constraints]

[]
  
[Executioner]
    type = Eigenvalue
    eigen_problem_type = gen_non_hermitian
    which_eigen_pairs = SMALLEST_MAGNITUDE
    n_eigen_pairs = 5
    n_basis_vectors = 18
    solve_type = jacobi_davidson
    petsc_options = '-eps_view'
[]
  
[Outputs]
    exodus = true
[]
```

**Reply by GiudGiud** [reply]
Thanks I ll try this out tomorrow. 
@cticenhour who is a good person for complex eigenvalue problems ?

**Reply by adhithya96** [reply]
Thanks @GiudGiud . I don't think this is a complex eigenvalue problem as we are using two different variables ur and uim to solve the problem.

**Reply by GiudGiud** [reply]
@adhithya96 Static1D does not compile for me. I had to change this:
```
  // Get point  locator
  const auto pl = _mesh.getPointLocator();
  // Locate node on the other side of the geometry
  // The translation vector will be a user parameter
  const auto current_point = *_current_node;
  //  const libMesh::Point current_point = pl->Point();
  const auto new_node = pl->locate_node(current_point);
```
in the BlochBC
also need to enable XFEM in the Makefile

**Reply by GiudGiud** [reply]
once it compiled I hit the same error as on the other discussion
```

*** ERROR ***
The following error occurred in the object "cut_mesh", of type "InterfaceMeshCut2DUserObject".

cutFragmentByGeometry not yet implemented for InterfaceMeshCut2DUserObject

```
this makes sense. I ll wait for your reply over there

**Reply by adhithya96** [reply]
My sincere apologies, 

I thought I had updated the BlochBC related files on github but I hadn't. I have updated the files. [eigen1d](https://github.com/adhithya96/EigenSolverMOOSE/tree/main). I have checked this. This one compiles properly. 
The overall problem I'm solving is to evaluate the  dispersion curves  for phononic crystals using XFEM. I have to solve the eigenvalue problem for multiple values of wave number. I have simplified the problem so that I can approach it one step at a time  
![BlochBCExpl](https://github.com/idaholab/moose/assets/89237852/d724cfa0-f680-4030-b6bc-59dce764e854)

**Reply by GiudGiud** [reply]
how are you getting around this?
```
*** ERROR ***
The following error occurred in the object "cut_mesh", of type "InterfaceMeshCut2DUserObject".

cutFragmentByGeometry not yet implemented for InterfaceMeshCut2DUserObject
```

I wanted to hit this error
```
Invalid NodalBC for eigenvalue problems, please use homogeneous (array) Dirichlet
```
but the other one came first

**Reply by adhithya96** [reply]
There are two folders in the GitHub link. Did you use the files in eigen1d ?

**Reply by adhithya96** [reply]
That has files for 1D eigenvalue problem. There is no xfem used in that

**Reply by GiudGiud** [reply]
ok you can just comment out this error created at line 420 of NonLinearEigenSystem.C in MOOSE.
The idea is to ban non-zero values dirichlet BCs for eigenvalue problems (because we are looking to find the homogeneous solution). But here your Dirichlet BCs are really meant to be periodic BCs so they should not be disallowed

**Reply by adhithya96** [reply]
Aah okay. Will do it tommorow

**Reply by adhithya96** [reply]
This was very helpful. Now I know a little bit more about how to debug. However, I ran into this error. 
```
*** ERROR ***
The following error occurred in the object "MOOSE Problem", of type "EigenProblem".

A libMesh::PetscSolverException was raised during FEProblemBase::computeJacobianTags
New nonzero at (0,0) caused a malloc
Use MatSetOption(A, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE) to turn off this check
```
I checked the MOOSE discussions [question](https://github.com/idaholab/moose/discussions/20164#discussion-3840458). Added this to the problem in the input file
```
    error_on_jacobian_nonzero_reallocation = false
```
I went throught the FEProblemBase error section
![petsc_error](https://github.com/idaholab/moose/assets/89237852/42e0b85b-e259-4cb4-b095-1842e438feb6)
Is there anything else I should be looking for ?

**Reply by GiudGiud** [reply]
I was going to suggest you do that.
What is happening is that the entries of the Jacobian are not the same between different nonlinear iterations. This is not expected, so something is 'conditionally' adding entries to the Jacobian somewhere in your code. Maybe in the material models.

**Reply by adhithya96** [reply]
I tried a few things
1. Changed the order of variables to ```SECOND```
2. Changed the solver used in eigenvalue problem. Tried ```Krylovschur``` and ```jacobi_davidson```
3. Chaged the material property from ```GenericConstantMaterial``` to elastic material with appropriate changes needed for tensor mechanics.
4. Changed the value of diffusivity to 0 and tried implementing the same code. 
All of these changes gives the same error. If the diffusivity is 0 then there shouldn't be any non-zero term getting added because of material model. So I think there is something wrong with the ```BlochDirichletBCImag``` and ```BlochDirichletReal```

**Reply by GiudGiud** [reply]
Do you get past this error?

**Reply by adhithya96** [reply]
No. I get the same error every time

**Reply by GiudGiud** [reply]
Ok that s normal. It won’t go away on its own

you need to comment out the lines I mentioned and recompile if we are talking about the first error

if about the second error, you need to pass that Boolean I mentioned to the problem

**Reply by adhithya96** [reply]
No which error are you talking about ? 
I'm talking about the PetSc error.

**Reply by adhithya96** [reply]
I did recompile after commenting it out. I didn't get that invalid NodalBC error again

**Reply by GiudGiud** [reply]
the petsc error you should be able to bypass with
```
    error_on_jacobian_nonzero_reallocation = false
```
in the Problem iirc. Do you have that?

**Reply by adhithya96** [reply]
Yes. I did. I am still getting the petsc error.

**Reply by adhithya96** [reply]
> This was very helpful. Now I know a little bit more about how to debug. However, I ran into this error. 
> 
> ```
> 
> *** ERROR ***
> 
> The following error occurred in the object "MOOSE Problem", of type "EigenProblem".
> 
> 
> 
> A libMesh::PetscSolverException was raised during FEProblemBase::computeJacobianTags
> 
> New nonzero at (0,0) caused a malloc
> 
> Use MatSetOption(A, MAT_NEW_NONZERO_ALLOCATION_ERR, PETSC_FALSE) to turn off this check
> 
> ```
> 
> I checked the MOOSE discussions [question](https://github.com/idaholab/moose/discussions/20164#discussion-3840458). Added this to the problem in the input file
> 
> ```
> 
>     error_on_jacobian_nonzero_reallocation = false
> 
> ```
> 
> I went throught the FEProblemBase error section
> 
> ![petsc_error](https://github.com/idaholab/moose/assets/89237852/42e0b85b-e259-4cb4-b095-1842e438feb6)
> 
> Is there anything else I should be looking for ?
> 
> 

I had tried it.

**Reply by GiudGiud** [reply]
can you please make a MWE? we need to investigate that on a simple case

**Reply by adhithya96** [reply]
I have put all the files that I am using in the github repo. [EigenSolverMOOSE](https://github.com/adhithya96/EigenSolverMOOSE). The files needed for this case is in ```eigen1d``` folder. 
By MWE I am hoping you minimum working example. I am using a 1D case with MatDiffusion kernel in the file  ```eigen1D_periodic_eigen.i```. I have used elastic  material properties in ```eigen1D_periodic_trial2.i```. This is the simplest example I could think of to start with. I have also uploaded the files of the moose version which I am using with the comments  added so that the following error doesn't happen. 
```Invalid NodalBC for eigenvalue problems, please use homogeneous (array) Dirichlet```

**Reply by adhithya96** [reply]
Also, I am not able find the element on the opposite side to implement the same case for a NeumannBC by modifying the functorNeumannBC. 
```
 error: no match for 'operator-' (operand types are 'const libMesh::Elem' and 'const libMesh::Point')
   59 |   const Elem * _new_elem = (*pl)(*_current_elem -  _translation_vec);
      |                                  ~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
      |                                  |                 |
      |                                  |                 const libMesh::Point
      |                                  const libMesh::Elem
 error: no match for 'operator-' (operand types are 'const libMesh::Elem' and 'const libMesh::Point')
   59 |   const Elem * _new_elem = (*pl)(*_current_elem -  _translation_vec);
      |                                  ~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
      |                                  |                 |
      |                                  |                 const libMesh::Point
      |                                  const libMesh::Elem
```
I went through the libmesh::Elem doxygen and the codes from XFEM.C and FEProblemBase.C to see if there is anything that I can use. I couldn't find anything. Any suggestions ?
This is my full code for BlochNeumannBC
```
 ADReal
 BlochNeumannBCReal::computeQpResidual()
 {
  //Get point  locator  
  const auto pl = _mesh.getPointLocator();

  //Locate node on the other side of the geometry
  //The translation vector will be a user parameter
  const auto _translation_vec= libMesh::Point(_lattice_vec, 0.0, 0.0);

  const auto _new_side = _current_side - _lattice_vec;

  const Elem * _new_elem = (*pl)(*_current_elem -  _translation_vec);

 if(!_new_elem)
 mooseError("Did not find the opposite side value");

 const Moose::ElemSideQpArg space_arg = {_new_elem, _new_side, _qp, _qrule, _q_point[_qp]};

 const Moose::StateArg time_arg = Moose::currentState();

 return _sign * _coef(space_arg, time_arg) * (_ur.gradient(space_arg, time_arg) * cos (_wave_num * _lattice_vec)  -  _uim.gradient(space_arg, time_arg) * sin (_wave_num * _lattice_vec)) * _test[_i][_qp];

 }
```

**Reply by GiudGiud** [reply]
you cant substract a point from an element. You can only subtract a point from a point
did you mean to do this?
_current_elem->vertex_average() - _translation_vec?

Or do you want to translate from the current quadrature point? ` _q_point[_qp]`

**Reply by adhithya96** [reply]
I need to find the new element and quadrature now that you mention it but to find '''space_arg'''.

**Reply by GiudGiud** [reply]
The quadrature point index is stil _qp imo
The element will be found by using `_current_elem->vertex_average() - _translation_vec` when locating an element with the point locator

**Reply by adhithya96** [reply]
That worked! Now I am getting this error. 
```
error: could not convert 'libMesh::TypeVector<T>::operator*(const Scalar&) const [with Scalar = double; T = MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>; typename libMesh::boostcopy::enable_if_c<libMesh::ScalarTraits<T2>::value, libMesh::TypeVector<typename libMesh::CompareTypes<T, T2>::supertype> >::type = libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true> >]((* &(&(&((BlochNeumannBCReal*)this)->BlochNeumannBCReal::<anonymous>.ADIntegratedBCTempl<double>::_test)->MooseArray<std::vector<double> >::operator[](((BlochNeumannBCReal*)this)->BlochNeumannBCReal::<anonymous>.ADIntegratedBCTempl<double>::<anonymous>.IntegratedBCBase::_i))->std::vector<double>::operator[](((std::vector<double>::size_type)((BlochNeumannBCReal*)this)->BlochNeumannBCReal::<anonymous>.ADIntegratedBCTempl<double>::<anonymous>.IntegratedBCBase::_qp))))' from 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true> > >::type' {aka 'libMesh::TypeVector<MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true> >'} to 'ADReal' {aka 'MetaPhysicL::DualNumber<double, MetaPhysicL::SemiDynamicSparseNumberArray<double, long unsigned int, MetaPhysicL::NWrapper<64> >, true>'}
   68 |  return _sign * _coef(space_arg, time_arg) * (_ur.gradient(space_arg, time_arg) * cos (_wave_num * _lattice_vec)  -  _uim.gradient(space_arg, time_arg) * sin (_wave_num * _lattice_vec)) * _test[_i][_qp];

```

**Reply by GiudGiud** [reply]
so the problem is that this whole expression is a vector, and it should be a scalar (it s the residual).
it's a vector because of these gradient terms
```
_ur.gradient(space_arg, time_arg)
```
Are you sure there is not supposed to be another vector that gets dot-producted with those gradients?
Should the gradients be there? Or should it just be one component of the gradient?
